### ISSUE TITLE
cram/writer: Add CRAM 3.1 support

### ISSUE THREAD

**Issue by zaeleus** [zaeleus/noodles#107]
This is a tracking issue for CRAM 3.1 write support.

The CRAM 3.1 format is still [a draft](https://github.com/samtools/hts-specs/pull/433), but [there is interest](https://github.com/zaeleus/noodles/issues/78#issuecomment-1220668241) in having noodles-cram demonstrate a second implementation.

The format is structurally the same as 3.0, but 3.1 includes additional block compression methods: rANS Nx16 (5), adaptive arithmetic coder (6), fqzcomp (7), and name tokenizer (8). noodles' CRAM reader supports decoding these compression methods already.

There is currently no way to apply user-defined compression methods for any particular data series. The implementation now simply uses gzip (DEFLATE) to compress all block data. As part of this work, a data series-compression method map must also be created to allow overriding the default encoders. Not only would this allow greater flexibility, but it would provide the framework to build selectable presets a la htslib's/samtools' [CRAM compression profiles](https://www.htslib.org/doc/samtools.html#GLOBAL_COMMAND_OPTIONS).

* [x] cram/writer/num/vlq: Add variable-length quantity (`uint7`) writer 
* [x] cram/codecs: Add rANS Nx16 encoder
  * [x] Order-0 encoder
  * [x] Order-1 encoder
  * [x] Stripe encoder
  * [x] RLE encoder
  * [x] Bit packing encoder
* [x] cram/codecs: Add adaptive arithmetic coding encoder
  * [x] Range coder encoder
  * [x] Statistical model encoder
  * [x] Order-0 encoder
  * [x] Order-1 encoder
  * [x] External (bzip2) encoder
  * [x] Stripe encoder
  * [x] Order-0 RLE encoder
  * [x] Order-1 RLE encoder
  * [x] Bit packing encoder
* [x] cram/codecs: Add fqzcomp encoder
* [x] cram/codecs: Add name tokenizer encoder
* [x] cram/codecs: Add encoder options
* [x] cram/data_container: Add block content-encoder map
* [x] cram/writer/builder: Allow overriding block compression methods for data series
* [x] cram/writer/builder: Select CRAM format version based on encoders
* [x] cram/codecs: Roundtrip validation
  * [x] rANS 4x8
  * [x] rANS Nx16
  * [x] adaptive arithmetic coding
  * [x] name tokenizer
  * [x] fqzcomp

**Comment by zaeleus** [comment]
All CRAM 3.1 (draft) encoders are now implemented (to at least MVP) and validated against htscodecs.

The `util_alignment_rewrite` example can be used for testing. The following shows how a `BlockContentEncoderMap` can set an encoder to specific block content. noodles-cram currently does not automatically select codecs based on content or have defined presets.

```diff
diff --git a/noodles-util/examples/util_alignment_rewrite.rs b/noodles-util/examples/util_alignment_rewrite.rs
index b9a656de..75328816 100644
--- a/noodles-util/examples/util_alignment_rewrite.rs
+++ b/noodles-util/examples/util_alignment_rewrite.rs
@@ -27,8 +27,34 @@ fn main() -> io::Result<()> {

     let header = reader.read_header()?;

+    use noodles_cram::{
+        codecs::{aac, rans_nx16, Encoder},
+        data_container::{
+            compression_header::{
+                data_series_encoding_map::DataSeries, preservation_map::tag_ids_dictionary::Key,
+            },
+            BlockContentEncoderMap,
+        },
+    };
+    use noodles_sam::record::data::field::{value::Type, Tag};
+
     let mut writer = alignment::writer::Builder::default()
         .set_reference_sequence_repository(repository)
+        .set_block_content_encoder_map(
+            BlockContentEncoderMap::builder()
+                .set_core_data_encoder(None)
+                .set_data_series_encoder(DataSeries::ReadNames, Some(Encoder::NameTokenizer))
+                .set_data_series_encoder(DataSeries::QualityScores, Some(Encoder::Fqzcomp))
+                .set_data_series_encoder(
+                    DataSeries::NumberOfReadFeatures,
+                    Some(Encoder::RansNx16(rans_nx16::Flags::empty())),
+                )
+                .set_tag_values_encoder(
+                    Key::new(Tag::AlignmentHitCount, Type::UInt8),
+                    Some(Encoder::AdaptiveArithmeticCoding(aac::Flags::empty())),
+                )
+                .build(),
+        )
         .build_from_path(dst)?;

     writer.write_header(&header)?;
```

To run, use

```
cargo run --release --all-features --example util_alignment_rewrite \
    <in.{sam,bam,cram}> \
    <out.cram> \
    [indexed-reference.fa]
```

noodles-cram also has CRAM 3.1 read support. To verify, use

```
cargo run --release --example cram_view <out.cram> [indexed-reference.fa]
```

or, to compare, use

```
samtools view [--reference <reference.fa>] <out.cram>
```

/cc @jkbonfield

**Comment by claymcleod** [comment]
Well done, @zaeleus 🎉👏

**Comment by ohofmann** [comment]
Oh wow. Indeed, yay!

**Comment by jkbonfield** [comment]
This is fantastic news.  Thanks for reporting the spec issues as you found them too.  It's been most useful.

The CRAM 3.1 spec is still draft, but having a 2nd implementation was one of the hurdles to cross which you have now done.  The second hurdle is just getting a pair of eyes to review it.  Obviously you've also done quite a bit of that too.  Did you have any suggestions for general improvements, over and above the silly typos and errors in it.

Note 3.1 focuses entirely on the codec layer, and not on the CRAM format itself.  I know there need to be changes there, such as switching everything to a better variable sized integer notation so we no longer have issues with 32-bit vs 64-bit quantities.  That's necessary to support chromosomes longer than 2GB.

**Comment by jkbonfield** [comment]
It's not clear how to test the writer.  I tried applying the patch above, but it didn't use fqzcomp.  io_lib's cram_size at the end reports:

```
Block CORE              , total size           0
Block content_id       1, total size     3917487 g                          BF
Block content_id       2, total size     1955049 g                          CF
Block content_id       4, total size      164722 g                          RL
Block content_id       5, total size     5605115 g                          AP
Block content_id       6, total size      110368 g                          RG
Block content_id       7, total size    47120001               n            RN
Block content_id       8, total size       32671 g                          MF
Block content_id       9, total size      171762 g                          NS
Block content_id      10, total size      965746 g                          NP
Block content_id      11, total size      288118 g                          TS
Block content_id      12, total size     4457323 g                          NF
Block content_id      13, total size      262593 g                          TL
Block content_id      14, total size      775586      0                     FN
Block content_id      15, total size     1139249 g                          FC
Block content_id      16, total size     2667873 g                          FP
Block content_id      17, total size       56591 g                          DL
Block content_id      18, total size    72639939 g                          BB
Block content_id      19, total size   153157930 g                          QQ
Block content_id      21, total size       78831 g                          IN
Block content_id      24, total size       96583 g                          HC
Block content_id      25, total size     7771953 g                          SC
Block content_id      26, total size      496863 g                          MQ
Block content_id      27, total size     1867986 g                          BA
Block content_id      28, total size     1209778 g                          QS
Block content_id 4281155, total size     5727616 g                          ASC
Block content_id 5063770, total size         697 g                          MDZ
Block content_id 5131587, total size         187 g                          NMC
Block content_id 5456218, total size     1257349 g                          SAZ
Block content_id 5783898, total size     5465827 g                          XAZ
Block content_id 5788483, total size     6474578 g                          XSC
```

So it got a couple new codecs, but is mostly gzip.  I can see fqzcomp listed in the patch, but most qualities appear to be generated using QQ instead of QS.  That's a legitimate strategy, although it dates back to the very first draft of CRAM pre 1.0 where the idea was to do lossy compression by only storing specific qualities (in interesting places) rather than the whole quality string.  Samtools can read it back fine though so it's still permitted.

**Comment by zaeleus** [comment]
By default, all block content encoders use gzip. This is due to not having presets. The ones in the patch explicitly override select block content.

Quality scores in features (lossy or not) and records with "quality scores stored as array" can interleave, so it's not obvious how fqzcomp is is supported in CRAM since each fqzcomp entry needs to know its length. In my implementation, fqzcomp has a constraint where all records in the slice must have the "quality scores stored as array" CRAM bit flag; otherwise, it falls back to gzip.

This means it's currently only applied to slices with only unmapped reads. For example,

```
@HD	VN:1.6
r1	4	*	0	255	*	*	0	0	ATCG	NDLS
r2	4	*	0	255	*	*	0	0	ATCG	OEMT
r3	4	*	0	255	*	*	0	0	ATCG	PFNU
```

Using the same patch as posted above:

```
$ cargo run --release --all-features --example util_alignment_rewrite in.sam out.cram
$ cram_size out.cram
Block CORE              , total size           0
Block content_id       1, total size          23 g                          BF
Block content_id       2, total size          23 g                          CF
Block content_id       4, total size          23 g                          RL
Block content_id       5, total size          23 g                          AP
Block content_id       6, total size          28 g                          RG
Block content_id       7, total size         145               n            RN
Block content_id       8, total size          23 g                          MF
Block content_id       9, total size          28 g                          NS
Block content_id      10, total size          23 g                          NP
Block content_id      11, total size          23 g                          TS
Block content_id      13, total size          23 g                          TL
Block content_id      27, total size          26 g                          BA
Block content_id      28, total size          34              f             QS
```

See that fqzcomp was used for `QS`.

> most qualities appear to be generated using QQ instead of QS. That's a legitimate strategy, although it dates back to the very first draft of CRAM pre 1.0

I don't think I've seen this described anywhere, so I had no idea. Are you saying that `QS` is preferred over `QQ` when the quality scores are lossless?

**Comment by jkbonfield** [comment]
> Quality scores in features (lossy or not) and records with "quality scores stored as array" can interleave, so it's not obvious how fqzcomp is is supported in CRAM since each fqzcomp entry needs to know its length. In my implementation, fqzcomp has a constraint where all records in the slice must have the "quality scores stored as array" CRAM bit flag; otherwise, it falls back to gzip.
> 
> This means it's currently only applied to slices with only unmapped reads. For example,

Thanks.  That gives me something to go on.


> > most qualities appear to be generated using QQ instead of QS. That's a legitimate strategy, although it dates back to the very first draft of CRAM pre 1.0
> 
> I don't think I've seen this described anywhere, so I had no idea. Are you saying that `QS` is preferred over `QQ` when the quality scores are lossless?

It's a mess frankly!  When I arrived on the CRAM scene we already had the distinction between QQ and QS, with cramtools able to generate both.  I'm not sure "preferred" is the right word, but it's what I went with in the absence of any other guidance.

Note it's also even more cumbersome than may initially appear.  If we have QS used by dint of setting the flag to say qualities are present, then any explicit quality value is in addition to the QS one, meaning we could potentially get conflicts with no clear definition of what takes precedence.  :/  I'm *guessing* QS would be applied first as a base layer and QQ etc gets added on as patches to it.  Maybe the theory was it could enable lossy encoding with crude smoothing and then precise per-base quality editing, but the history as to why it's like it is was lost.

The issue of fqzcomp and knowing the read-length is an interesting one too.  There are odd situations where the QS record may be longer than the sequence, such as when we get an ambiguity code in reference and/or sequence.  This requires the `B` feature (or since CRAM 3.0 `b` is possible, which maybe I should switch to using), which for some unknown reason has a strict requirement of both adding to bases (BA) and qualities (QS), even though the quality is already known.  This then gets seq len and QS record out of sync.  However fqzcomp can embed the quality string length in the fqz data stream so it can handle these foibles.  I just validated this by taking 10,000 fixed length sequences and created a copy of the SAM with one ambiguity code added.  The former CRAM had length stored once, while the latter had lengths stored per quality string to accommodate the +1 byte.  The QS size grew by 0.03% so it barely had an impact in that case.

**Comment by jkbonfield** [comment]
I can confirm I can get it to generate fqzcomp data.  Samtools decodes it fine, and obviously as already checked the vice versa is true.

However I did find a file that noodles was unable to re-read back.  This was its own generated file.  Samtools can decode it fine, so it's maybe a noodles decode issue rather than encode one.  That seemed related to the presence of `GO:query` in the input SAM header.

```
@HD	VN:1.6	SO:unsorted	GO:query
VP2-06:112:H7LNDMCVY:1:1273:1687:24048	77	*	0	0	*	*0	0	TTGCCCAGGCTGGTCTTGAATTCACGGGCTCAAGCCATCCTCCTAGCTGAGCCTCCTGAGTAGCTGGGATTACAGGTGTGAGCCATTGCACCCAACTGGTATAGCCACGTTAGAAAACAGTCTGGCAGTTTCTCAAAAGGCTAAATGTACA	EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE3EEEEEEEEEEE3EEEEEE3EEEEEEEEEEEEEEEEEEEEEEEEEEEEEE3EEEEEEEEEEEEEEEE3EEEEEEEEEEEEEEEEEE3E3EEE3-E3-E3EE
```

I can add this as a separate issue if you prefer.  Removing the GO:query fixes this round trip.

**Comment by zaeleus** [comment]
@jkbonfield, thanks for taking the time to test this.

The history of CRAM is also valuable knowledge. You can see that I took the straightforward approach to guarantee the collection of quality score lengths. There's certainly room for improvement in the future.

> However I did find a file that noodles was unable to re-read back.

It was indeed a SAM header issue with the group order field. It's tracked in #122.

**Comment by zaeleus** [comment]
The specification for CRAM 3.1 was published today in samtools/hts-specs@7efd20489f746d6b335f3283ac9546211674ba71. I'm marking this issue as done since the implementation remains complete.

As per recommendation in § 6 "[File definition](https://samtools.github.io/hts-specs/CRAMv3.pdf#section.6)", noodles will continue to write CRAM 3.0 by default unless CRAM 3.1 codecs are used.

**Comment by jkbonfield** [comment]
Firstly, a huge thank you @zaeleus for doing this.  We couldn't have got it out the door without your work.

The rest is only indirectly related to CRAM 3.1.   For CRAM 3.2 or CRAM 4.0, how responsive do you think you can be for future spec changes?

I deliberately limited what went into CRAM 3.1 as it was already clear progress was slowing down in the traditional "other camp" for CRAM support, but there are definitely things that need changing.  There are some simpler changes also like adding Zstd support could help with unmapped data, but I wanted to avoid a kitchen-sink scenario unless there was substantial evidence for need (this is gathering momentum though).

There are other more invasive things such as removing the 32-bit liimt on a lot of quantities is a road block for some organisms with long chromosomes.

In short, I basically stopped CRAM spec updates because it was clear adding more would simply make adoption of a newer spec an even more distant affair.  However there are still things on the roadmap to consider.  Collaboration would be very welcome, both implementation obviously but also most definitely in the ideas and strategies.  Feel free to contact me directly if you wish (jkb at sanger ac uk).
